### DISCUSSION TITLE
Stop (GUI) server in Linux ?

### DISCUSSION THREAD

**Discussion by pcar75** [jamulussoftware/jamulus#3006] [Technical Support & Help] [answered]
Bonjour,
On my dual boot PC, I can shutdown Jamulus server in Windows 10.  However I can't do that in Linux, because I believe it installs itself as a service.  Is there a nicer (GUI) way of shuting down Jamulus server, instead of running a script or terminal with  `killall -I "jamulus"` ?

Context : dual Boot PC : Windows 10 / Ubuntu Studio 22.04.1 LTS (Plasma KDE).
On Windows
    Jamulus server (3.9.1) with GUI
        Menu > Window > Exit (Ctrl+Q) : OK terminates process & threads
        X Button : OK terminates process & threads
On Ubuntu Studio 22.04.1 LTS
    Jamulus server (3.9.1) with GUI and Icon in system tray bar
        Menu > Window > Exit (Ctrl+Q) : Hides GUI only
        X (Close) Button : Hides GUI only
        Sys tray Exit (Icon menu buttons = Exit, Show, Hide...) : Hides GUI if it was visible ( ~ Exit ICON menu).

EDIT : added precision : Ubuntu Studio 22.04.1 LTS and Qt is 5.15.3 as mentioned below (rdica).

**Comment by rdica** [comment]
I installed the deb on my Ubuntu Studio 22.04 DAW and I dont see a systemd service file. Their are two GUI shortcuts, one starts the client, one starts the server with `jamulus -s`. There's no service to stop once you've stopped the GUI server.

Interestingly enough, and I think it may be a bug, but I cannot stop the GUI server client. It seems to ignore exit. I had to `killall jamulus` from CLI.

**Reply by pljones** [reply]
> the GUI server client

What does this mean?

**Reply by rdica** [reply]
The server. Exiting the GUI does not stop the server.

**Reply by pljones** [reply]
Strange.  I don't get that behaviour with 3.91 GUI server on Linux.

**Reply by rdica** [reply]
Which Linux?

**Reply by rdica** [reply]
For the record, Ubuntu Studio 22.04 uses QT version 5.15.3.

**Reply by rdica** [reply]
Tried compiling from src on Ubuntu Studio 22.04 and get same result, cannot stop server process when trying to exit GUI.

**Reply by pljones** [reply]
I'm running KDE.  I've a deep dislike of GNOME (mostly as it won't run over remote X).
```
$ uname -a
Linux fs-peter 5.15.0-60-lowlatency #66-Ubuntu SMP PREEMPT Tue Jan 24 14:26:16 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
$ gcc --version
gcc (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
$ apt list libqt5gui5
libqt5gui5/jammy-updates,now 5.15.3+dfsg-2ubuntu0.2 amd64 [installed,automatic]
```
```
$ git checkout main
Switched to branch 'main'
Your branch is up-to-date with 'origin/main'.
$ git pull upstream main
From github.com:jamulussoftware/jamulus
 * branch              main       -> FETCH_HEAD
Already up-to-date.
$ make distclean; qmake && make -j3
...
g++ -Wl,--as-needed -Wl,-O1 -o Jamulus x86cpu.o ... moc_clientrpc.o  -ljack -lpthread /usr/lib/x86_64-linux-gnu/libQt5Widgets.so /usr/lib/x86_64-linux-gnu/libQt5Multimedia.so /usr/lib/x86_64-linux-gnu/libQt5Gui.so /usr/lib/x86_64-linux-gnu/libQt5Network.so /usr/lib/x86_64-linux-gnu/libQt5Xml.so /usr/lib/x86_64-linux-gnu/libQt5Concurrent.so /usr/lib/x86_64-linux-gnu/libQt5Core.so -lGL
```
```
$ ./Jamulus -p 60001 -s
- selected port number: 60001
- server mode chosen
Using "X.Y.Z.A" as external IP.
Ignoring invalid serverinfo, please verify the parameter syntax.
Recording state not initialised
Recording state: enabled
... (any option to close: Ctrl-Q or Window->Exit (with mouse or keyboard), or window Close button, or Alt-F4)
$
```
```
$ ps -ef | grep 60001
login-id 128784  127928  0 09:02 pts/0    00:00:00 grep --color=auto 60001
```
So no process left over.

**Reply by rdica** [reply]
Ubuntu Studio 22.04 is KDE based, they changed from XFCE at some point after 18.04.

I tried exiting with Ctrl-Q or Window->Exit (with mouse or keyboard), or window Close button, or Alt-F4, both with my compiled binary and the deb pkg binary, and process could only be stopped with CLI to kill the process.

I attached strace to the running process before trying to exit the GUI but I didnt see anything in the log that indicated a problem.

Any other suggestions I could try for troubleshooting?

**Reply by pljones** [reply]
You could try the following to get some more diagnostics.  In `src/serverdlg.cpp`:
- after line 36, add 
```
std::cout << qUtf8Printable ( QString ( "CServerDlg: bSystemTrayIconAvailable=%s\n").arg( bSystemTrayIconAvailable ) ) );
```
- after line 494, add
```
std::cout << qUtf8Printable ( QString ( "CServerDlg::closeEvent: event received\n" ) );
```
- after line 894, add
```
std::cout << qUtf8Printable ( QString ( "CServerDlg::changeEvent: bSystemTrayIconAvailable=true; pEvent->type=WindowStateChange; isMinimized=%s\n").arg( isMinimized() ) ) );
```

**Reply by rdica** [reply]
To be clear, I'm editing 3.9.1 release, not some other branch, correct?
I get compile errors adding the lines you gave me:
![compile](https://user-images.githubusercontent.com/16609801/218331206-f1fc13ea-a49a-46f8-a42a-1fbbdf01a25f.png)

**Reply by ann0see** [reply]
Instead of std::cout use qDebug()

**Reply by rdica** [reply]
It would be far easier if you could provide a patch. I'm getting errors using qDebug() in place of std::cout.

**Reply by rdica** [reply]
![compile2](https://user-images.githubusercontent.com/16609801/218522812-550db4b9-4e24-44a6-ab5d-526369a75e7f.png)

**Reply by pljones** [reply]
> It would be far easier if you could provide a patch. I'm getting errors using qDebug() in place of std::cout.

`patch -p1` then paste this in should work:
```
diff --git a/src/serverdlg.cpp b/src/serverdlg.cpp
index 874c0814..8ea4e91c 100644
--- a/src/serverdlg.cpp
+++ b/src/serverdlg.cpp
@@ -34,6 +34,7 @@ CServerDlg::CServerDlg ( CServer* pNServP, CServerSettings* pNSetP, const bool b
 {
     // check if system tray icon can be used
     bSystemTrayIconAvailable = SystemTrayIcon.isSystemTrayAvailable();
+qInfo() << QString ( "CServerDlg: bSystemTrayIconAvailable=%1 " ).arg ( bSystemTrayIconAvailable  );

     setupUi ( this );

@@ -493,6 +494,7 @@ CServerDlg::CServerDlg ( CServer* pNServP, CServerSettings* pNSetP, const bool b

 void CServerDlg::closeEvent ( QCloseEvent* Event )
 {
+qInfo() << "CServerDlg::closeEvent: event received";
     // store window positions
     pSettings->vecWindowPosMain = saveGeometry();

@@ -892,6 +894,7 @@ void CServerDlg::changeEvent ( QEvent* pEvent )
     // minimized
     if ( bSystemTrayIconAvailable && ( pEvent->type() == QEvent::WindowStateChange ) )
     {
+qInfo() << QString ( "CServerDlg::changeEvent: bSystemTrayIconAvailable=true; pEvent->type=WindowStateChange; isMinimized=%1" ).arg ( isMinimized() );
         if ( isMinimized() )
         {
             // we have to call the hide function from another thread -> use
```

**Reply by rdica** [reply]
I tried each different way to close the window, and as shown in the image, the process is still running.

![compile3](https://user-images.githubusercontent.com/16609801/218556164-57678a01-d95c-4454-95a8-a342bc2227fa.png)

Contents of `~/.config/Jamulus/Jamulusserver.ini`:

```
<server>
<winposmain_base64>AdnQywADAAAAAAR9AAAAOwAABsgAAALBAAAEfQAAADsAAAbIAAACwQAAAAAAAAAAB4AAAAR9AAAAOwAABsgAAALB</winposmain_base64>
 <directorytype>1</directorytype>
 <name>Server GUI Cant Stop</name>
 <city>Midwest</city>
 <country>225</country>
 <norecord>0</norecord>
 <welcome>VWJ1bnR1IFN0dWRpbyAyMi4wNA==</welcome>
 <language>en</language>
 <recordingdir_base64>L2hvbWUvYXJkeS9qYW11bHVzX3Byb2plY3Rz</recordingdir_base64>
 <directoryaddress></directoryaddress>
 <directoryfile_base64></directoryfile_base64>
 <autostartmin>0</autostartmin>
 <delaypan>1</delaypan>
</server>
```

**Reply by pljones** [reply]
Mmmm...  `bSystemTrayIconAvailable=1` means that there should be a visible System Tray icon.  If not, Qt is lying to us...  Can you check your KDE desktop configuration that you've not got something hidden but active where the tray would be (I don't have a KDE desktop handy)?  If so, does the icon appear when you run Jamulus Server?  (Note that killing the application with Ctrl-C may not remove the icon.)

Windows:
![image](https://user-images.githubusercontent.com/1549463/218562759-41b85321-9d71-4f74-addb-8ed1d65bda5a.png)
... and right-click on it:
![image](https://user-images.githubusercontent.com/1549463/218563859-0dd6ff12-8455-4836-a55b-c5df13ee225e.png)
where "Show" and "Hide" should trigger the appropriate `pEvent->type=WindowStateChange` diagnostic and "Exit" should exit.

**Reply by pljones** [reply]
The `event received` line means that Qt is telling Jamulus the user has asked to close the app.  Jamulus says "OK go ahead".  Again, it's down to Qt to actually end the application.

**Reply by rdica** [reply]
The icon tray item is visible and was one of the means of exiting I tried.
Show/Hide do not generate any messages. Exit generates one as shown above, but doesn't actually exit the program. All means of exiting DO close the GUI window, but the icon is still in the tray, and process is still running. When I ctrl+c in the terminal I started jamulus in, then the icon disappears, process ends, I'm returned to command prompt.

**Reply by rdica** [reply]
![compile5](https://user-images.githubusercontent.com/16609801/218568095-cca2c578-8015-494e-be31-a6df224dc857.png)

**Reply by pljones** [reply]
Thanks for confirming.

**Reply by rdica** [reply]
Fired up an instance of Xubuntu (XFCE4) 22.04, installed 3.9.1 deb, ran the server from menu and also from CLI, and in both cases the server exited correctly.

**Reply by pljones** [reply]
Thanks for trying that, too.  Good to hear.  That does narrow things down.

**Reply by rdica** [reply]
Spun up a Kubuntu 22.04 instance (not Ubuntu Studio) and saw the same behavior shown in Ubuntu Studio.

**Comment by pljones** [comment]
So far, no one else has commented running Linux but _not_ running **Ubuntu Studio 22.04 LTS**.  I can't think what Ubuntu Studio would have done but it would be good to hear other experiences.  Could others with this issue please comment.

**Comment by pcar75** [comment]
So, to my understanding , the problem seems to be related to KDE vs Qt vs applications' setup and use of these ... (not Ubuntu).  
An initial web search points to somewhat similar problems (centered around KDE).  I will continue to search the web for more / better info.

Meanwhile, more details ...
@rdica Can you confirm the same specs ?

uname -a
      Linux user-kubuntu-pc 5.15.0-60-generic #66-Ubuntu SMP Fri Jan 20 14:29:49 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

lsb_release -a
      No LSB modules are available.
      Distributor ID: Ubuntu
      Description:    Ubuntu 22.04.1 LTS
      Release:        22.04
      Codename:       jammy

In destop's System Settings > About this System
    Operating System: Kubuntu 22.04
    KDE Plasma Version: 5.24.4
    KDE Frameworks Version: 5.92.0
    Qt Version: 5.15.3
    Kernel Version: 5.15.0-60-generic (64-bit)
    Graphics Platform: X11
    (...)

N.b.: I installed Ubuntu Studio 22.04 LTS packaging over Kubuntu 22.04 LTS.
I did follow a specific install referenced in a video on Ubuntu Studio* page, that can allow a low-latency kernel and can "grub-boot" it.  However, having seen no major difference between generic and low-latency for jamulus sessions , I normally boot to generic kernel.

*Mauro Gaspari video tutorial with
https://frontpagelinux.com/tutorials/easily-install-jack-audio-with-ubuntu-studio-installer/
(except nothing special for Carla as I did not plan on using it)  

- Enable realtime process priority

and
Select packages to install :

- linux-lowlatency
- ubuntustudio-lowlatency-settings
- ubuntustudio-performance-tweaks

**Reply by rdica** [reply]
The kernel versions have no bearing. On the non-Ubuntu Studio installs I used the default generic kernel AND the lowlatency kernel.

I suspect KDE/kwin as pkg'd by Ubuntu, still thinking how I can isolate and further test/troubleshoot...

**Reply by rdica** [reply]
I spun up the Fedora 37 KDE spin, installed build deps, and compiled 3.9.1 from src.
QT 5.15.8
gcc 12.2.1
KDE 22.12.2

Same behaviour!
Fedora seems to default to Wayland, I saw one message in the terminal I started Jamulus from, and it was caused by clicking 
 "Show Jamulus server" from tray icon.

![compile6](https://user-images.githubusercontent.com/16609801/219502178-952b510f-d622-4bb0-8068-79b37949462a.png)

Had to ctrl+c in the terminal to terminate the process.

**Reply by rdica** [reply]
I forgot to note, my installation of Ubuntu Studio 22.04 uses Xorg, not Wayland.

**Comment by pcar75** [comment]
I posted this questions in Ask Ubuntu, to further search for a solution.
https://askubuntu.com/questions/1455310/cannot-terminate-jamulus-server-under-kde

**Comment by pcar75** [comment]
From a few observations ...

- rdica's trace : "CServerDlg: bSystemTrayIconAvailable=1"

- pljones : "The event received line means that Qt is telling Jamulus the user has asked to close the app. Jamulus says "OK go ahead". Again, it's down to Qt to actually end the application.

- There is an application called "Discover" that also has an icon in the system tray.  When I menu-cliked it to Exit, a pop-up appeared to verify that I really wanted to Exit this program.  This has never happened with the Jamulus server system tray icon's Exit.

Is it possible to test a version that does not install it's "icon/etc." in the system tray ?  Just to see if this could rule out a now different behavior between KDE and Qt.  Sorry, but I'm not Linux-savy enough to install and test it on my system without breaking something.

**Reply by pljones** [reply]
src/serverdlg.cpp around line 35, change
```
bSystemTrayIconAvailable = SystemTrayIcon.isSystemTrayAvailable();
```
to
```
bSystemTrayIconAvailable = false; //SystemTrayIcon.isSystemTrayAvailable();
```
then rebuild with `make distclean; qmake && make` and run with `./Jamulus -s`

**Reply by rdica** [reply]
I recompiled using the above, lo and behold, exiting actually worked. There was no tray icon, which was expected.
ctrl-q and alt-F4 both worked as well as exiting via the GUI.

**Comment by pcar75** [comment]
@rdica Thanks for testing my hypothesis :-)

@pljones I think I should make this into a bug report. Do you agree ?

**Reply by pljones** [reply]
It's not a bug in Jamulus.  You can log a feature request to make it possible to turn off iconise to system tray. :)

(Me, I'd rather the feature was totally removed, but I'm like that.)

**Comment by rdica** [comment, marked as answer]
I filed an issue from this discussion.

**Reply by pcar75** [reply]
@rdica Thanks !  To me, I would rather classify it as a bug or at least undesirable behavior of the system tray icon ...
Issue = Current KDE prevents Server GUI from exiting #3016 
Closing this one.
